### PR TITLE
Improve and add fixer for  prefer-template

### LINKED ISSUE #3034: prefer-template: Add config to restrict series of string concatinations
It looks like https://github.com/palantir/tslint/pull/2741 allows a series of string concatenations, which I'd like to disable in our codebase. The specific use case was someone using the `+` operator to concatenate a string with a template literal, when it would be preferred to use a single template literal. Since it looks like this was an intentional change, I'd like to add configuration to `prefer-template` to work around this.

simple proposal
```
{
  "type": "array",
  "enum": [
    "allow-single-concat",
    "allow-pure-string-concat" // enables the change introduced in #2741
  ]
}
```

### Bug Report

- __TSLint version__: 5.4.3 (introduced in 5.3.0)
- __TypeScript version__: 2.3.4
- __Running TSLint via__: any

#### TypeScript code being linted

```ts
const world = "world";

console.log("hello" + " world"); // a
console.log("hello" + ` ${world}`); // b
console.log("hello " + world); // c
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "prefer-template": true
    }
}
```

#### Actual behavior
a: no warning
b: no warning
c: warning

#### Expected behavior
a: no warning
b: warning
c: warning

### PR DESCRIPTION
**Pull request by smoehrle**
#### PR checklist

- [x] Addresses an existing issue: #3034 
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:

- Made rule stricter
  - String concatenation is only allowed if the expressions does not contain a newline and there is at most one expression per line. This fixes #3034.
- Add fixer
  - Currently only for the single line case

#### CHANGELOG.md entry:

[new-fixer] prefer-template
[bugfix] string concatenation on the same line is an error now

### CODE REVIEW

**Comment by scott-ho** [comment]
@suchanlee can you have a look over this PR? It seems good. The the CI problem seems not to be the pr's problem.

**Comment by suchanlee** [comment]
Merging master to the branch or rebasing off of master should fix the circle build

**Comment by scott-ho** [comment]
@smoehrle could you perform a rebase?

**Comment by smoehrle** [comment]
Done

**Comment by blossom-park** [comment]
when would fixer for prefer-template option added? I want fixer for prefer-template option asap. ESLint already has it

File: src/rules/preferTemplateRule.ts [hunk 1]
```diff
@@ -59,42 +73,68 @@ export class Rule extends Lint.Rules.AbstractRule {
...
+        }
+
+        const evalExpressions = evaluateExpressions(expressions, ctx.sourceFile);
+
+        if (evalExpressions.some((exp) => exp.isStringLike && exp.hasNewLine)) {
+            // Multiline template
+            // Currently no fixer available
+            ctx.addFailureAtNode(node, Rule.FAILURE_STRING_MULTILINE);
+        } else {
+            // (Concatenated) single line stinglikes
```
Comment (JoshuaKGoldberg) [inline src/rules/preferTemplateRule.ts]: > stinglkes

Typo: should be `stringlikes`.

File: src/rules/preferTemplateRule.ts [hunk 2]
```diff
@@ -27,11 +27,24 @@ interface Options {
...
 export class Rule extends Lint.Rules.AbstractRule {
     /* tslint:disable:object-literal-sort-keys */
     public static metadata: Lint.IRuleMetadata = {
         ruleName: "prefer-template",
-        description: "Prefer a template expression over string literal concatenation.",
+        description: Lint.Utils.dedent`
+            Prefer a template expression over string literal concatenation.
+
+            There is one expection for this rule regarding long strings. It is allowed
+            to concatenate stings and template expressions if they do not contain a
```
Comment (JoshuaKGoldberg) [inline src/rules/preferTemplateRule.ts]: > stings

Typo: should be `strings`

**Comment by ericanderson** [comment]
Because this changes a rule semantic, its a breaking change.

**Comment by macrozone** [comment]
@adidahiya eslint has a working prefer-template fix implementation, that plays nicely with prettier. So maybe look how they solve it.

**Comment by JoshuaKGoldberg** [comment]
Given #4534 and pending status on requested changes, I'm going to close this for housekeeping. Anybody: so long as #3034 is open, please do send a new PR if you'd like to push these changes forward!

**Pull request closed without merging**